src/modules/cpu/kernel/log.hpp

```diff
@@ -0,0 +1,131 @@
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "rppdefs.h"
+#include "rpp_cpu_common.hpp"
+
+// 1 pixel log helper functions
+inline void compute_log(Rpp8s *src, Rpp32f *dst) { *dst = log(abs((Rpp32f)(*src + 128))); }
+inline void compute_log(Rpp16f *src, Rpp16f *dst) { *dst = log(abs((Rpp16f)(*src))); }
+template<typename T>
+inline void compute_log(T *src, Rpp32f *dst) { *dst = log(abs((Rpp32f)(*src))); }
+
+// Computes ND log recursively
+template<typename T1, typename T2>
+void log_recursive(T1 *src, Rpp32u *srcStrides, T2 *dst, Rpp32u *dstStrides, Rpp32u *dstShape, Rpp32u nDim)
+{
+    if (!nDim)
+    {
+        compute_log(src, dst);
+    }
+    else
+    {
+        for (int i = 0; i < *dstShape; i++)
+        {
+            log_recursive(src, srcStrides + 1, dst, dstStrides + 1, dstShape + 1, nDim - 1);
+            dst += *dstStrides;
+            src += *srcStrides;
+        }
+    }
+}
+
+template<typename T1, typename T2>
+RppStatus log_generic_host_tensor(T1 *srcPtr,
+                                  RpptGenericDescPtr srcGenericDescPtr,
+                                  T2 *dstPtr,
+                                  RpptGenericDescPtr dstGenericDescPtr,
+                                  Rpp32u *roiTensor,
+                                  rpp::Handle& handle)
+{
+    Rpp32u numThreads = handle.GetNumThreads();
+    Rpp32u nDim = srcGenericDescPtr->numDims - 1; // Omitting batchSize here to get tensor dimension.
+    Rpp32u batchSize = dstGenericDescPtr->dims[0];
+
+    omp_set_dynamic(0);
+#pragma omp parallel for num_threads(numThreads)
+    for(int batchCount = 0; batchCount < batchSize; batchCount++)
+    {
+        Rpp32u *roi = roiTensor + batchCount * nDim * 2;
+        Rpp32u *begin = roi;
+        Rpp32u *length = &roi[nDim];
+
+        T1 *srcPtr1 = srcPtr + batchCount * srcGenericDescPtr->strides[0];
+        T2 *dstPtr1 = dstPtr + batchCount * dstGenericDescPtr->strides[0];
+
+        for(int i = 0; i < nDim; i++)
+            srcPtr1 += begin[i] * srcGenericDescPtr->strides[i + 1];
+        if(nDim == 2)
+        {
+            T1 *srcPtrRow = srcPtr1;
+            T2 *dstPtrRow = dstPtr1;
+
+            for(int i = 0; i < length[0]; i++)
+            {
+                T1 *srcPtrTemp = srcPtrRow;
+                T2 *dstPtrTemp = dstPtrRow;
+
+                for (int vectorLoopCount = 0; vectorLoopCount < length[1]; vectorLoopCount++)
+                {
+                    compute_log(srcPtrTemp, dstPtrTemp);
```

Comment:
Also how do we handle cases when input is 0?
log(0) is undefined. So please explore online methods to handle such cases

Reply:
For now i am using comparing max(abs_output, 1) to avoid undefined.

Reply:
where is this comparison done?

Reply:
Please let me know on your thoughts.

Reply:
In DALI, for log(0) they return error stating expects input between 1 and 64.
In python, for log(0) they return -inf.

I will use std::nextafter(0.0f, 1.0f) as discussed with Sampath.